### PR TITLE
enhancement: register all proposals of specific nodes

### PR DESCRIPTION
**Pull request by maxwellmattryan**
## Summary

> Please summarize your changes, describing **what** they are and **why** they were made.

We need to add logic to query and register events from given nodes.

:warning: Do NOT merge until #5964 has been merged!

## Changelog

```
- Please write a list of changes
```

### Platforms

> Please select any platforms where your changes have been tested.

-   **Desktop**
    -   [ ] MacOS
    -   [x] Linux
    -   [ ] Windows
-   **Mobile**
    -   [ ] iOS
    -   [ ] Android

### Instructions

> Please describe the specific instructions, configurations, and/or test cases necessary to **test and verify** that your changes work as intended.

...

## Checklist

> Please tick the following boxes that are relevant to your changes.

-   [x] I have followed the contribution guidelines for this project
-   [x] I have performed a self-review of my own code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have added or modified tests that prove my changes work as intended
-   [ ] I have verified that new and existing unit tests pass locally with my changes
-   [ ] I have verified that my latest changes pass CI workflows for testing and linting
-   [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: packages/desktop/components/popups/AddNodePopup.svelte [hunk 1]
```diff
@@ -5,12 +5,14 @@
...
     import { showAppNotification } from '@auxiliary/notification'
+    import { Platform } from 'shared/lib/core/app'
+    import { registerProposalsFromNode } from 'shared/lib/contexts/governance'
 
     export let node: INode = EMPTY_NODE
     export let isEditingNode: boolean = false
     export let onSuccess: (..._: any[]) => void
 
-    const currentNode = node
+    const currentNode = { ...node }
```
Comment (MarkNerdi) [inline packages/desktop/components/popups/AddNodePopup.svelte:15]: do you do this to copy the object without reference? 
if yes, you would need a deep copy, eg. `JSON.parse(JSON.stringify())`

Comment (maxwellmattryan) [reply packages/desktop/components/popups/AddNodePopup.svelte:15]: Forgot that we have a `deepCopy` util function that does this in `core/utils/object.ts`

**Comment by nicole-obrien** [comment]
This is not adding all events when a user creates a new wallet/account. We can add a new issue for that.

File: packages/shared/components/modals/NodeActionsMenu.svelte [hunk 2]
```diff
@@ -56,7 +56,7 @@
         } else {
             await togglePrimaryNodeInClientOptions(node)
             if (Platform.isFeatureFlagEnabled('governance')) {
-                await registerProposalsFromPrimaryNode()
+                await registerProposalsFromNode(node)
```
Comment (nicole-obrien) [inline packages/shared/components/modals/NodeActionsMenu.svelte:59]: If we are now registering all proposals at login and also when we add a new node, we do not need to register a proposal at changing a primary node.

File: packages/desktop/components/popups/AddProposalPopup.svelte [hunk 3]
```diff
@@ -74,7 +74,11 @@
 
     async function registerParticipationWrapper(auth?: Auth): Promise<void> {
         const accounts = toAllAccounts ? $activeAccounts : [$selectedAccount]
-        const promises = accounts.map((account) => registerParticipationEvent(eventId, { url: nodeUrl, auth }, account))
+        const options = {
+            node: { url: nodeUrl, auth },
+            eventsToRegister: [eventId],
+        }
+        const promises = accounts.map((account) => registerParticipationEvents(options, account))
```
Comment (nicole-obrien) [inline packages/desktop/components/popups/AddProposalPopup.svelte:81]: This should be an action

File: packages/desktop/components/popups/AddNodePopup.svelte [hunk 4]
```diff
@@ -30,6 +33,13 @@
             } else {
                 await addNodeToClientOptions(node)
             }
+
+            if (Platform.isFeatureFlagEnabled('governance')) {
+                if (currentNode?.url !== node?.url) {
+                    await registerProposalsFromNode(node)
```
Comment (nicole-obrien) [inline packages/desktop/components/popups/AddNodePopup.svelte:39]: Why are we not doing it for all accounts?

File: packages/shared/lib/core/profile/actions/active-profile/login.ts [hunk 5]
```diff
@@ -143,7 +144,7 @@ export async function login(loginOptions?: ILoginOptions): Promise<void> {
             void pollMarketPrices()
             if (Platform.isFeatureFlagEnabled('governance')) {
                 void initializeRegisteredProposals()
-                void registerProposalsFromPrimaryNode()
+                void registerProposalsFromNodes(clientOptions?.nodes ?? [])
```
Comment (nicole-obrien) [inline packages/shared/lib/core/profile/actions/active-profile/login.ts:147]: Why are we not registering them for all accounts?

File: packages/shared/lib/contexts/governance/actions/registerProposalsFromNode.ts [hunk 6]
```diff
@@ -3,22 +3,11 @@ import { get } from 'svelte/store'
...
-
-    const _selectedAccount = get(selectedAccount)
-    await Promise.all(
-        proposalIds.map(async (proposalId) => {
-            if (!isProposalAlreadyAddedForSelectedAccount(proposalId)) {
-                await registerParticipationEvent(proposalId, node, _selectedAccount)
-            }
-        })
-    )
+    await registerParticipationEvents(options, get(selectedAccount))
```
Comment (nicole-obrien) [inline packages/shared/lib/contexts/governance/actions/registerProposalsFromNode.ts:12]: Why is this for the selected account? We will need to register events for any account.

**Review by nicole-obrien — CHANGES_REQUESTED** [review]

File: packages/desktop/components/popups/AddNodePopup.svelte [hunk 7]
```diff
@@ -30,6 +33,13 @@
             } else {
                 await addNodeToClientOptions(node)
             }
+
+            if (Platform.isFeatureFlagEnabled('governance')) {
+                if (currentNode?.url !== node?.url) {
```
Comment (nicole-obrien) [inline packages/desktop/components/popups/AddNodePopup.svelte:38]: Why do we check if its the current node, if a node has been added or edited then we should pull in the events

File: packages/shared/lib/core/account/actions/tryCreateAdditionalAccount.ts [hunk 8]
```diff
@@ -1,15 +1,25 @@
...
 
 export async function tryCreateAdditionalAccount(alias: string, color: string): Promise<void> {
     try {
         const account = await createNewAccount(alias, color)
         setSelectedAccount(account?.index)
+
+        if (Platform.isFeatureFlagEnabled('governance')) {
+            const { nodes } = get(activeProfile)?.clientOptions
+            await Promise.all(nodes.map((node) => registerParticipationEvents({ node }, account)))
+        }
```
Comment (nicole-obrien) [inline packages/shared/lib/core/account/actions/tryCreateAdditionalAccount.ts:21]: Should this be an action?

File: packages/shared/lib/core/profile/actions/active-profile/login.ts [hunk 9]
```diff
@@ -143,7 +144,9 @@ export async function login(loginOptions?: ILoginOptions): Promise<void> {
             void pollMarketPrices()
             if (Platform.isFeatureFlagEnabled('governance')) {
                 void initializeRegisteredProposals()
-                void registerProposalsFromPrimaryNode()
+                void Promise.all(
+                    clientOptions?.nodes.map((node) => registerProposalsForAccounts({ node }, get(activeAccounts)))
```
Comment (nicole-obrien) [inline packages/shared/lib/core/profile/actions/active-profile/login.ts:148]: Can this be an action?

**Review by nicole-obrien — CHANGES_REQUESTED** [review]

**Review by nicole-obrien — APPROVED** [review]
LGTM!

**Pull request merged**